### PR TITLE
Reasoning competition

### PR DESCRIPTION
**Pull request by RodrigoLPA**
Pending To-Dos:

- [ ] Agree on release block number
- [ ] Make sure that 3B and potentially 16k-token-long traces is reasonable
- [ ] Test on testnet. 
- [ ] Review competition parameters.
- [ ] Test diff number of samples to evaluate
- [ ] Make sure that loss ceiling still makes sense in `method.py`

### CODE REVIEW

File: finetune/datasets/hugging_face/hugging_face_loader.py [hunk 1]
```diff
@@ -251,26 +259,507 @@ def fetch_dataset_configs(self) -> typing.Dict[str, typing.Dict]:
...
+        if tokenizer.pad_token is None:
+            tokenizer.pad_token = tokenizer.eos_token
+        
+        for q, t, a, tt in zip(self.questions, self.traces, self.answers, self.task_types):
+            # Only include samples from supported task types
+            if tt not in self.supported_task_types:
+                continue
+            
+            # Tokenize question
+            question_tokens = tokenizer.encode(
```
Comment (cryptal-mc) [inline finetune/datasets/hugging_face/hugging_face_loader.py:687]: A potential issue here is that no instruct chat template is applied. Instruct models are usually queries after apply a chat template to the list of messages (the messages column in the dataset).
Here is an example of how chat templates are applied in LLama.
https://www.llama.com/docs/model-cards-and-prompt-formats/llama3_1/

Also here is an example from the macrobench dataloader for multiple choice questions:

https://github.com/macrocosm-os/finetuning/blob/2d8980e5cf79e61c7ee015f1d9f61774be32848e/finetune/datasets/hugging_face/macrocosmos_dataset_loader.py#L183

**Review by cryptal-mc — CHANGES_REQUESTED** [review]

File: finetune/datasets/hugging_face/hugging_face_loader.py [hunk 2]
```diff
@@ -251,26 +259,507 @@ def fetch_dataset_configs(self) -> typing.Dict[str, typing.Dict]:
...
-            batches.append(
-                np.array([input_ids + [tokenizer.eos_token_id]])
+    ) -> typing.Union[typing.List[np.ndarray], typing.Dict[str, typing.List]]:
+        """Tokenize the dataset based on what's needed.
+        
+        For text_loss evaluation, returns list of tokenized samples.
+        For verifiable_reasoning evaluation, returns dictionary with structured data.
+        """
+        # Get the caller's stack frame to determine which evaluation method is calling
+        caller_frame = inspect.currentframe().f_back
```
[the remark this answers is not in the document]
Comment (RodrigoLPA) [reply finetune/datasets/hugging_face/hugging_face_loader.py:269]: Agreed, I added the eval_method as an argument for tokenize()
